modules/common/service/service.go

```diff
@@ -67,42 +108,143 @@ func (s *Service) GetServiceHostname() string {
 	return s.serviceHostname
 }
 
-// GetServiceHostnamePort - returns the service hostname with port
-func (s *Service) GetServiceHostnamePort() string {
-	return fmt.Sprintf("%s:%d", s.GetServiceHostname(), GetServicesPortDetails(s.service, s.service.Name).Port)
+// GetServiceHostnamePort - returns the service hostname with port if service port
+// is not nil, otherwise returns GetServiceHostname()
+func (s *Service) GetServiceHostnamePort() (string, string) {
+	servicePort := GetServicesPortDetails(s.service, s.service.Name)
+	if servicePort != nil {
+		return s.GetServiceHostname(),
+			strconv.FormatInt(int64(GetServicesPortDetails(s.service, s.service.Name).Port), 10)
+	}
+
+	return s.GetServiceHostname(), ""
+}
+
+// GetLabels - returns labels of the service
+func (s *Service) GetLabels() map[string]string {
+	return s.service.Labels
+}
+
+// GetAnnotations - returns annotations of the service
+func (s *Service) GetAnnotations() map[string]string {
+	return s.service.Annotations
+}
+
+// GetSpec - returns the spec of the service
+func (s *Service) GetSpec() *corev1.ServiceSpec {
+	spec := s.service.Spec
+	return &spec
+}
+
+// GetServiceType - returns type of the service spec
+func (s *Service) GetServiceType() corev1.ServiceType {
+	return s.service.Spec.Type
 }
 
 // AddAnnotation - Adds annotation and merges it with the current set
 func (s *Service) AddAnnotation(anno map[string]string) {
 	s.service.Annotations = util.MergeStringMaps(s.service.Annotations, anno)
 }
 
+// GetAPIEndpoint - returns the API endpoint URL for the service to register in keystone.
+func (s *Service) GetAPIEndpoint(svcOverride *OverrideSpec, protocol *Protocol, path string) (string, error) {
+	var apiEndpoint *url.URL
+	var err error
+	if svcOverride != nil && svcOverride.EndpointURL != nil {
+		apiEndpoint, err = url.Parse(*svcOverride.EndpointURL)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		hostname, port := s.GetServiceHostnamePort()
+
+		var endptURL string
+		if protocol != nil &&
+			((*protocol == ProtocolHTTP && port == "80") ||
+				(*protocol == ProtocolHTTPS && port == "443")) {
+			endptURL = fmt.Sprintf("%s%s", EndptProtocol(protocol), hostname)
+		} else {
+			endptURL = fmt.Sprintf("%s%s:%s", EndptProtocol(protocol), hostname, port)
+		}
+
+		// Update instance status with service endpoint url from route host information
```

Comment:
Is this an old copy-pasted comment?  We're not really updating the instance status here, right?

Reply:
yes thats an outdated comment. will fix that, thanks!